client/lib/url-search/use-url-search.ts

```diff
@@ -0,0 +1,79 @@
+import { useState } from '@wordpress/element';
+import { addQueryArgs } from '@wordpress/url';
+import debugFactory from 'debug';
+import page from 'page';
+
+const debug = debugFactory( 'calypso:url-search' );
+
+/**
+ * Function for constructing the url to page to. Here are some examples:
+ * 1. { uri: 'google.com', search:'hello' } --> 'google.com?s=hello'
+ * 2. {
+ *     uri: 'wordpress.com/read/search?q=reader+is+awesome',
+ *     search: 'reader is super awesome'
+ *     queryKey: 'q',
+ *    } --> 'wordpress.com/read/search?q=reader+is+super+awesome'
+ *
+ * @param {object} options the options object
+ * @param {string} options.uri the base uri to modify and add a query to
+ * @param {string} options.search the search term
+ * @param {string} [options.queryKey = s] the key to place in the url.  defaults to s
+ * @returns {string} The built search url
+ */
+export const buildSearchUrl = ( {
+	uri,
+	search,
+	queryKey = 's',
+}: {
+	uri: string;
+	search?: string;
+	queryKey?: string;
+} ) => {
+	const query = addQueryArgs( uri, { [ queryKey ]: search } );
+
+	if ( ! search ) {
+		const queryStringIndex = uri.indexOf( '?' );
+
+		if ( queryStringIndex !== -1 ) {
+			return uri.substring( 0, queryStringIndex );
+		}
+
+		return uri;
+	}
+
+	return query.replace( /%20/g, '+' );
+};
+
+function useUrlSearch( queryKey = 's' ) {
+	const [ isSearchOpen, setSearchOpen ] = useState( false );
+
+	function doSearch( query: string ) {
+		setSearchOpen( '' !== query );
+
+		const searchURL = buildSearchUrl( {
+			uri: window.location.href,
+			search: query,
+			queryKey: queryKey,
```

Comment:
Something I wanted changed in the original implementation, could we replace this with a generic "queryParams" array and move away from the idea that search is a single phrase?

Most search implementations need filters. I'm wanting to drop marketplace categories through here at a later time.

Reply:
While this is a good idea, I wonder if it would still be a "search" implementation. It feels like it would rather be a "filter" implementation, which feels like a departure from this PR. I definitely don't see this as a requirement for this PR.

Reply:
Interesting @lsl, could you provide one example of where it will be used? So, it will be easy for to me see how can I implement it in the best way I can.

Reply:
@tyxla I'm thinking of [URLSearchParams](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) when I think of search.

@paulopmt1 Gabriel tried to add `queryParams` to this hook a sprint or two ago, we ended up going with the existing category implementation in the marketplace as a temporary solution because it already existed. However that implementation has a lot of problems with regards to where we want to take marketplace search and I was planning to remove those routes and replace it with query params - by replacing this hook and adding query params to it.

Happy to ship this as is for now but lets get queryParams in there before we move the other usages over to the new hook to avoid doubling testing.